### PR TITLE
YUNIKORN-418: Add "config" REST API

### PR DESCRIPTION
**Pull request by manirajv06**
First cut implementation. In addition, response middleware has been added to return error also in json format and used httptest package for unit testing the same.

### CODE REVIEW

File: pkg/webservice/handlers.go [hunk 1]
```diff
@@ -457,7 +457,36 @@ func getClusterConfig(w http.ResponseWriter, r *http.Request) {
...
+	queryParams := r.URL.Query()
+	_, dryRunExists := queryParams["dry_run"]
+	if len(queryParams) == 0 || !dryRunExists {
+		http.Error(w, "Mandatory parameters are missing in URL path. Please check the usage documentation", http.StatusBadRequest)
+		return
+	}
+	if queryParams.Get("dry_run") != "1" {
+		http.Error(w, "Invalid query params. Please check the usage documentation", http.StatusBadRequest)
+		return
+	}
```
Comment (holdenk) [inline pkg/webservice/handlers.go:472]: This seems to require that dry run is set, which is confusing. Could the error message maybe state what parameter is missing or what parameter is invalid?

File: pkg/webservice/webservice.go [hunk 2]
```diff
@@ -58,14 +59,40 @@ func newRouter() *mux.Router {
...
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		start := time.Now()
-
-		inner.ServeHTTP(w, r)
-
+		rw := &YResponseWriter{ResponseWriter: w, statusCode: -1}
+		inner.ServeHTTP(rw, r)
+		var body = string(rw.body)
+		if rw.statusCode != 200 && rw.statusCode != -1 {
+			errorInfo := dao.NewYAPIError(nil, rw.statusCode, body)
```
Comment (kingamarton) [inline pkg/webservice/webservice.go:66]: This `NewYAPIError` method is not defined. This is the reason why Travis check is failing.

File: pkg/webservice/handlers.go [hunk 3]
```diff
@@ -457,7 +457,36 @@ func getClusterConfig(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
-func updateConfig(w http.ResponseWriter, r *http.Request) {
+func createClusterConfig(w http.ResponseWriter, r *http.Request) {
+	writeHeaders(w)
+
+	queryParams := r.URL.Query()
+	_, dryRunExists := queryParams["dry_run"]
```
Comment (kingamarton) [inline pkg/webservice/handlers.go:464]: Please don't ignore the returned value here and reuse it in the line 469:

```
dryRun, dryRunExists := queryParams["dry_run"]
....
if dryRun != "1" {
...
```

File: pkg/webservice/handlers.go (hunk 1 above)
Comment (kingamarton) [reply pkg/webservice/handlers.go:472]: Right now we support only dry_run=1. I think we should document it and also include it in the error message.

**Review by kingamarton — CHANGES_REQUESTED** [review]

File: pkg/webservice/handlers.go (hunk 1 above)
Comment (manirajv06) [reply pkg/webservice/handlers.go]: Taken care. Filed a separate JIRA - https://issues.apache.org/jira/browse/YUNIKORN-515 to track doc/user guide changes.

File: pkg/webservice/routes.go [hunk 4]
```diff
@@ -102,7 +102,15 @@ var webRoutes = routes{
 		"Scheduler",
 		"PUT",
 		"/ws/v1/config",
-		updateConfig,
+		updateClusterConfig,
+	},
+
+	// endpoint to update the current conf
```
[the remark this answers is not in the document]
Comment (manirajv06) [reply pkg/webservice/routes.go:108]: Yes, changes are mainly for validation through createClusterConfig method. For update, just renamed "updateConfig" to "updateClusterConfig"

File: pkg/webservice/webservice.go (hunk 2 above)
Comment (manirajv06) [reply pkg/webservice/webservice.go:66]: Missed this file. Added now.

File: pkg/webservice/handlers.go (hunk 3 above)
Comment (manirajv06) [reply pkg/webservice/handlers.go]: Taken care

File: pkg/webservice/dao/error_info.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+package dao
+
+type YAPIError struct {
+	StatusCode  int    `json:"status_code"`
+	Message     string `json:"message"`
+	Description string `json:"description"`
+}
+
+func NewYAPIError(err error, statusCode int, message string) *YAPIError {
```
Comment (kingamarton) [inline pkg/webservice/dao/error_info.go:27]: The `err` parameter is not used. The description should have the value of the `err error` parameter, no?

File: pkg/webservice/routes.go (hunk 4 above)
Comment (kingamarton) [inline pkg/webservice/routes.go:108]: please update the comment here, since this is the endpoint for validating the config, not for updating it.

**Review by kingamarton — CHANGES_REQUESTED** [review]

File: pkg/webservice/dao/error_info.go (hunk 5 above)
Comment (manirajv06) [reply pkg/webservice/dao/error_info.go:27]: Taken care

File: pkg/webservice/routes.go (hunk 4 above)
Comment (manirajv06) [reply pkg/webservice/routes.go]: Taken care

**Review by kingamarton — APPROVED** [review]

**Pull request closed without merging**